File: GoScootIoS/GoScootIoS/Repository/AuthenticationRepository.swift
//
//  AuthenticationRepository.swift
//  GoScootIoS
//
//  Created by KumikoOumae on 8/11/25.
//

import Foundation
class AuthenticationRepository {
    
    func signIn(
        body: Request_LogInDTO,
    ) async throws -> Response_LogInDTO {
        let apiConfig = ApiConfig(apiPath: .logIn)
        var request = apiConfig.buildRequest(params: [], sessionId: nil)
        request.httpBody = try JSONEncoder().encode(body)
        let (data, resp) = try await URLSession.shared.data(for: request)
        guard let http = resp as? HTTPURLResponse else {
            throw ApiError(message: "Vừa xảy ra lỗi. Xin vui lòng thử lại", status: -1)
        }
        if (200..<300).contains(http.statusCode) {
            return try JSONDecoder().decode(Response_LogInDTO.self, from: data)
        } else if (http.statusCode == 401 || http.statusCode == 403){
            let error = try JSONDecoder().decode(Response_FailureDTO.self, from: data)
            throw UnAuthorizedError(message: error.message, status: http.statusCode)
        }
        else {
            let error = try JSONDecoder().decode(Response_FailureDTO.self, from: data)
            throw ApiError(message: error.message, status: http.statusCode)
        }
        
    }
    
    func signUp(
        body: Request_RegistrationDTO,
    ) async throws -> Void {
        let apiConfig = ApiConfig(apiPath: .register)
        var request = apiConfig.buildRequest(params: [], sessionId: nil)
        request.httpBody = try JSONEncoder().encode(body)
        let (data, resp) = try await URLSession.shared.data(for: request)
        guard let http = resp as? HTTPURLResponse else {
            throw ApiError(message: "Vừa xảy ra lỗi. Xin vui lòng thử lại", status: -1)
        }
        if (200..<300).contains(http.statusCode) {
            return 
        } else if (http.statusCode == 401 || http.statusCode == 403){
            let error = try JSONDecoder().decode(Response_FailureDTO.self, from: data)
            throw UnAuthorizedError(message: error.message, status: http.statusCode)
        }
        else {
            let error = try JSONDecoder().decode(Response_FailureDTO.self, from: data)
            throw ApiError(message: error.message, status: http.statusCode)
        }
        
    }
    
    func logInWithSessionId(
        sessionId: String,
    ) async throws -> Response_LogInDTO {
        let apiConfig = ApiConfig(apiPath: .logInViaSession)
        let request = apiConfig.buildRequest(params: [], sessionId: sessionId)
        
        let (data, resp) = try await URLSession.shared.data(for: request)
        guard let http = resp as? HTTPURLResponse
        else {
            throw ApiError(message: "Vừa xảy ra lỗi. Xin vui lòng thử lại", status: -1)
        }
        print(http.statusCode)
        if (200..<300).contains(http.statusCode) {
            return try JSONDecoder().decode(Response_LogInDTO.self, from: data)
        }
        else if (http.statusCode == 401 || http.statusCode == 403 || http.statusCode == 409){
            let error = try JSONDecoder().decode(Response_FailureDTO.self, from: data)
            throw UnAuthorizedError(message: error.message, status: http.statusCode)
        }
        else {
            let error = try JSONDecoder().decode(Response_FailureDTO.self, from: data)
            throw ApiError(message: error.message, status: http.statusCode)
        }
        
    }
}

File: GoScootAndroid/app/src/main/java/com/example/goscootandroid/Repository/Retrofit/AuthenticationRepository.kt
package com.example.goscootandroid.Repository

import android.annotation.SuppressLint
import android.util.Log
import com.example.goscootandroid.EnvironmentObjects.GoScootApi
import com.example.goscootandroid.Models.DTOs.Requests.RequestLogInDTO
import com.example.goscootandroid.Models.DTOs.Requests.RequestRegistrationDTO
import com.example.goscootandroid.Models.DTOs.Responses.ResponseFailureDTO
import com.example.goscootandroid.Models.DTOs.Responses.ResponseLogInDTO

import kotlinx.serialization.json.Json
import okhttp3.ResponseBody
import retrofit2.HttpException
import javax.inject.Inject

/**
 * Swift parity:
 * - signIn(body) -> ResponseLogInDTO
 * - signUp(body) -> Unit
 * - logInWithSessionId(sessionId) -> ResponseLogInDTO
 *
 * Throws:
 * - UnAuthorizedError on 401/403/409
 * - ApiError on other non-2xx, with server message if available
 */
class AuthenticationRepository @Inject constructor(
    private val api: GoScootApi,
    private val json: Json
) {

    suspend fun signIn(body: RequestLogInDTO): ResponseLogInDTO =
        try {
            Log.d("Sign In", "Signing In....")
            api.logIn(body)
        } catch (t: Throwable) {
            throw mapHttpException(t)
        }

    suspend fun signUp(body: RequestRegistrationDTO) {
        try {
            api.register(body)
        } catch (t: Throwable) {
            throw mapHttpException(t)
        }
    }

    suspend fun logInWithSessionId(sessionId: String): ResponseLogInDTO =
        try {
            api.logInViaSession(sessionId)
        } catch (t: Throwable) {
            throw mapHttpException(t)
        }

    // ---- helpers ----

    private fun mapHttpException(t: Throwable): Throwable {
        if (t is HttpException) {
            val code = t.code()
            val message = parseServerMessage(t.response()?.errorBody())

            return when (code) {
                401, 403, 409 -> UnAuthorizedError(message ?: "Unauthorized", code)
                else          -> ApiError(message ?: "HTTP $code", code)
            }
        }
        return ApiError(t.message ?: "Network error", -1)
    }

    private fun parseServerMessage(body: ResponseBody?): String? {
        if (body == null) return null
        return runCatching {
            val raw = body.string() // consume once
            json.decodeFromString<ResponseFailureDTO>(raw).message
        }.getOrNull()
    }
}